src/main/java/com/vmware/vipclient/i18n/messages/service/ComponentService.java

```diff
@@ -31,41 +32,48 @@ public ComponentService(MessagesDTO dto) {
     }
 
     /*
-     * get messages from local bundle or from remote vip service(non-Javadoc)
+     * Get messages from local bundle or from remote vip service(non-Javadoc)
      * 
      * @see
      * com.vmware.vipclient.i18n.messages.service.IComponentService#getMessages
      * (com.vmware.vipclient.i18n.base.DataSourceEnum)
      */
     @SuppressWarnings("unchecked")
-    public Map<String, String> getMessages() {
+    public Map<String, String> getMessages(final Map<String, Object> cacheProps) {
         Map<String, String> transMap = new HashMap<String, String>();
+        ComponentBasedOpt opt = new ComponentBasedOpt(dto);
         if (VIPCfg.getInstance().getMessageOrigin() == DataSourceEnum.VIP) {
-            transMap = new ComponentBasedOpt(dto).getComponentMessages();
+        	Map<String, Object> response = opt.getComponentMessages(cacheProps);
+	    	transMap = opt.getMsgsJson(response);
+	    	cacheProps.clear();
+	    	cacheProps.putAll(response);
         } else if (VIPCfg.getInstance().getMessageOrigin() == DataSourceEnum.Bundle) {
             transMap = new LocalMessagesOpt(dto).getComponentMessages();
         }
         return transMap;
     }
 
     public Map<String, String> getComponentTranslation() {
-        Map<String, String> retMap = new HashMap<String, String>();
         CacheService cs = new CacheService(dto);
-        retMap = cs.getCacheOfComponent();
-        if (retMap == null
-                && VIPCfg.getInstance().getCacheMode() == CacheMode.DISK) {
+        CacheItem cacheItem = cs.getCacheOfComponent(); 
+        cacheItem = cacheItem == null ? new CacheItem() : cacheItem;
+        Map<String, String> cachedMessages = cacheItem.getCachedData();
+ 
+        if (cachedMessages.isEmpty() && VIPCfg.getInstance().getCacheMode() == CacheMode.DISK) {
             Loader loader = VIPCfg.getInstance().getCacheManager()
                     .getLoaderInstance(DiskCacheLoader.class);
-            retMap = loader.load(dto.getCompositStrAsCacheKey());
+            cachedMessages = loader.load(dto.getCompositStrAsCacheKey());
         }
-        if (retMap == null && !cs.isContainComponent()) {
-            Object o = this.getMessages();
+        if (cachedMessages.isEmpty() && !cs.isContainComponent()) {
+        	Map<String, Object> cacheProps = new HashMap<String, Object>();
+            Object o = this.getMessages(cacheProps);
             Map<String, String> dataMap = (o == null ? null
                     : (Map<String, String>) o);
-            cs.addCacheOfComponent(dataMap);
-            retMap = dataMap;
+            
+            cs.addCacheOfComponent(new CacheItem (dataMap, cacheProps));
```

Comment:
Let's say there's a product having 100 components, how many 'cacheProps' instances would be in the memory?

Reply:
I think you didn't get my points.
Firstly it doesn't matter with the 'pre-OOP' at all. It's just a concept for 'container' which may be just a simple hashmap to store the etag with component-key, it's separated from MessageCache.
Secondly, looks like you don't get my question about 100 components. Per your change, i think there will be 100 cacheProps(map) instances in the memory which doesn't work if there are more components, in other words it limits the support to large components case.

Reply:
Colin, I think you are the one not getting my point.

See if this helps:
Each CacheProps map has the following key value pairs.
"RESPONSE_TIME" : "xxxxxx"
"HEADER" : "list of headers"
We  store this map inside CacheItem.

Now if you store this in one "container"... let's say a huge HashMap,
the key will still be the cacheKey, and the value will be this map above. You will still store 100 maps, each map associated to the key. The only difference is your putting the cacheProps map in another big hashmap "container" instead of individually, in each CacheItem. Do you see my point now?

Reply:
If you still doubt , I have tested implementations with 100 cache keys, 1000 messages each. I compared both implementation's memory consumption. See that using a separate container uses slightly higher heap memory (perhaps due to references)


<img width="1069" alt="Using Separate Container class for cacheProps" src="https://user-images.githubusercontent.com/5719531/78827948-cc6e6c00-7998-11ea-8971-50c7d2dcd7c6.png">
<img width="1066" alt="Using CacheItem cacheProps" src="https://user-images.githubusercontent.com/5719531/78827955-d001f300-7998-11ea-9ffb-5d7776811822.png">
<img width="633" alt="Test Code" src="https://user-images.githubusercontent.com/5719531/78827957-d09a8980-7998-11ea-9c8d-fdb1acc255e7.png">

Reply:
What's the reason to store HEADER, RESPONSE_CODE, RESPONSE_MESSAGE? I think they not required to keep in memory; instead, we only need to store etag and response timestamp? so, why we can't just use one map with component as key, etag-timestamp as value? With this, the map size is the number of components, not really huge.
Overall, the principle should keep the memory concise, maintainable and controllable.

Reply:
We can remove response code and response message, but it is the same concept. the key has to be component-locale, the existing cacheKey in MessageCache. Each item in the data cache has a corresponding set of etag/timestamp... you will store this in a separate Map.. as I did above. Same memory consumption, but 2 steps to store/query in different places, 
Frankly, concatenating etag and timestamp adds processing overhead!

And what happens if you add more properties later on you create another map? Very poor design, Colin, Please review OOP concepts.

Reply:
Since there's agreement on reducing cacheProps instances, mark it to resolve.